src/include/daos_cont.h

```diff
@@ -138,6 +138,35 @@ int
 daos_cont_create(daos_handle_t poh, const uuid_t uuid, daos_prop_t *cont_prop,
 		 daos_event_t *ev);
 
+/**
+ * Create a new container with label \a label on the storage pool connected
+ * by \a poh.
+ *
+ * \param[in]	poh	Pool connection handle.
+ * \param[in]	label	Required, label property of the new container.
+ *			Supersedes any label specified in \a cont_prop.
+ * \param[in]	cont_prop
+ *			Optional, container properties pointer
+ *			that if specified must not include an entry
+ *			with type DAOS_PROP_CO_LABEL.
+ * \param[out]	uuid	Optional, pointer to uuid_t to hold the
+ *		        implementation-generated container UUID.
+ * \param[in]	ev	Completion event, it is optional and can be NULL.
+ *			The function will run in blocking mode if \a ev is NULL.
+ *
+ * \return		These values will be returned by \a ev::ev_error in
+ *			non-blocking mode:
+ *			0		Success
+ *			-DER_INVAL	Invalid parameter
+ *			-DER_NO_HDL	Invalid pool handle
+ *			-DER_NO_PERM	Permission denied
+ *			-DER_UNREACH	network is unreachable
+ */
+int
+daos_cont_create_by_label(daos_handle_t poh, const char *label,
+			  daos_prop_t *cont_prop, uuid_t *uuid,
```

Comment:
Should we change

    uuid_t *uuid

to

    uuid_t uuid

? See man page uuid_generate(3). Apparently uuid_t is a pointer at the API level, and its implementation is also a pointer. ("uuid_t *" is sort of more difficult to program with.)

Reply:
Good catch.  uuid_t seems to be a odd to work with because of the way it's defined.

Reply:
I had it defined with uuid_t type initially in the patch. In UX WG a request was made to make it possible for a caller who does not care about the generated UUID to be able to pass NULL as an argument parameter in this case. 

Having the argument as uuid_t I think then requires all callers to have a uuid_t even if it may not be used subsequent to calling this function.

Reply:
So there are 3 possibilities. I think I am gravitating toward choice 3 below - let me see how that might be workable. @johannlombardi thoughts?

1) restore argument type to uuid_t
2) keep argument type as uuid_t *
3) remove the argument altogether - in which case if a container is created by label, and subsequently a user wants to discover the UUID for some unforeseen reason, another API call can be made, e.g., daos_cont_open_by_label() that returns daos_cont_info_t

Reply:
my vote would be 2.
 - it's optional so passing NULL would work for users who don't care.
 - for callers who want the uuid, there is no need to call 3 APIs to get it.

Reply:
> > > i'm maybe missing something here, but if one creates a task:
> > > DAOS_OPC_CONT_DESTROY
> > > and does not set the uuid, but sets the label in daos_cont_destroy_t;
> > > where does that get interpreted?
> > > dc_cont_destroy does not handle a label here right since you created a separate dc_cont_destroy_lbl?
> > > hmm im afraid this might be the same issue with open_by_label that we've missed. so am i missing something?
> > 
> > 
> > oh for open we probably do handle it with the internal open check:
> > cont_op = args->label ? CONT_OPEN_BYLABEL : CONT_OPEN;
> > but i don't see a similar logic for destroy?
> 
> actually looking at this more, task open is still wrong here because in dc_cont_open we check:
> if (uuid_is_null(args->uuid) || args->coh == NULL) and fail.. so label would not work.
> so I think that's an issue for both open and destroy

I had implemented the code such that dc_cont_destroy() does not deal with labels, and dc_cont_destory_lbl does not deal with container UUIDs. Much like the associated daos_cont_destroy() and daos_cont_destroy_by_label() functions, not wanting to change API signature of daos_cont_destroy().
dc_cont_destroy() sends the existing cart RPC DAOS_CONT_DESTROY
dc_cont_destroy_lbl() sends the new cart RPC DAOS_CONT_DESTROY_BYLABEL
Both RPCs have identical response message type, and can be handled by the existing completion function that does not need to distinguish the label vs. UUID case.

open was done roughly the same way (two different task functions meant specifically for each usage UUID or label). In the implementation, a little more complicated, resulting in a common dc_cont_open_internal() function for the code that sends an RPC (DAOS_CONT_OPEN or DAOS_CONT_OPEN_BYLABEL). And also had to distinguish label vs. UUID case in the RPC completion function (by label response gives the UUID information).

does this make sense? does the code violate design principles of the task interface?

Reply:
> my vote would be 2.
> 
> * it's optional so passing NULL would work for users who don't care.
> * for callers who want the uuid, there is no need to call 3 APIs to get it.

Fine with me.